cpp/src/arrow/util/thread_pool.h

```diff
@@ -393,8 +420,23 @@ class ARROW_EXPORT SerialExecutor : public Executor {
     RunLoop();
     return final_fut;
   }
+
+#ifndef ARROW_ENABLE_THREADING
+  // we have to run tasks from all live executors
+  // during RunLoop if we don't have threading
+  static std::unordered_set<SerialExecutor*> all_executors;
```

Comment:
Because CPU processing can wait for i/o. And in theory (not sure but I think in practice in acero) people can create multiple executors with dependencies across each other, which should still work in this case, just with things in other executors being run when one executor waits for the other.